### ISSUE TITLE
Beta 3: ERROR:flutter/lib/ui/painting/codec.cc(89) when loading some JPG files (worked fine in Beta 2)

### ISSUE THREAD

**Issue by freewheelnat** [flutter/flutter#17430]
My app includes video chat via Twilio on Android and takes some screenshots, which are saved to Firebase Storage. The screenshots are then shown on the home screen of the app.

With Beta 2, everything worked fine. But since upgrading to Beta 3, I sometimes get this error:
```
E/flutter ( 3414): [ERROR:flutter/lib/ui/painting/codec.cc(89)] Failed decoding image. Data is either invalid, or it is encoded using an unsupported format.
I/flutter ( 3414): ══╡ EXCEPTION CAUGHT BY SERVICES ╞══════════════════════════════════════════════════════════════════
I/flutter ( 3414): The following _Exception was thrown resolving an image codec:
I/flutter ( 3414): Exception: operation failed
```

This is for jpg files that previously could be shown without any issue with Beta 2. 

The jpg files are loaded as Uint8List from Firebase storage into `myImageData` then, to show them, I use 
```
new Image.memory(
          myImageData,
          fit: BoxFit.cover)
```

JPG that throws error:

![doesntwork](https://user-images.githubusercontent.com/4062052/39819009-3590fc1a-539a-11e8-8da3-7cf5d0eafe41.jpg)

This JPG works fine:
![works](https://user-images.githubusercontent.com/4062052/39819010-35b02e64-539a-11e8-8204-7379cf6ea181.jpg)

Both JPGs above showed fine in the app, with the exact same code, before upgrading to Beta 3.

(Edited for typo)

**Comment by jason-simmons** [comment]
I can render these images in the current master branch engine using `Image.asset`, or by loading them into a `Uint8List` from the asset bundle and using `Image.memory`

Do you have a standalone app that reproduces this issue?

**Comment by freewheelnat** [comment]
OK, I'll try it on master branch to see if it works (I'm using beta branch). I'll do that on Monday and update this issue then.

**Comment by freewheelnat** [comment]
OK, this is a really puzzling one. 

Indeed, the "error" image loads fine, whether I load it from assets, or from a url on my website, on Beta 3.

So then I thought issue was in firebase, but it doesn't appear to be.

First, let me explain in a few more details what my app does. It is a chat app. For each contact, there is a stack of messages. Some messages are text, some are videos. For videos, the app shows screenshots. If there is more than 1 screenshot available for a given video, the app loops through the screenshots. The looping is done with an animation, screenshot changes every 2 seconds. The widget showing the screenshot(s) is its own widget, with the animation in it, so only that widget is redrawn, not the whole screen.

So, the test I did was, instead of loading the real screenshots, I amended the firebase query to always load from the same path. I chose the path for the error jpg.

To my surprise, sometimes it rendered, sometimes it didn't!

So I thought, is firebase not loading it properly consistently? The screenshot, although the same, is loaded for each Video as all I did was direct firebase to the same path, instead of the path of the real Video screenshot(s). So I added some simple logs as below, with data being the Uint8List loaded from Firebase:

```
        print('image ' + data.length.toString());
        print('image ' + data.elementSizeInBytes.toString());
        print('image ' + data.elementAt(0).toString());
        print('image ' + data.elementAt(1).toString());
        print('image ' + data.elementAt(2).toString());
        print('image ' + data.elementAt(3).toString());
        print('image ' + data.elementAt(4).toString());
        print('image ' + data.elementAt(5).toString());
        print('image ' + data.elementAt(6).toString());
        print('image ' + data.elementAt(7).toString());

```

For each screenshot loaded from firebase, I got the same data.

```
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

```

OK, so it would seem like the screenshot is loading properly from firebase each time.

So then, I added the same logs to the UintList in the widget build method.

Now, something weird happened...

```
I/flutter (31165): video -L7y4yAQAwN0f_YSmntg 7gJ0lLcqgjfXtdu37wxj indexScreenshotToShow 0 available 2
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

I/flutter (31165): video -L7y4yAQAwN0f_YSmntg hIFyv3PUxXliHRwkKa5k indexScreenshotToShow 0 available 7
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

I/flutter (31165): video -L7y4yAQAwN0f_YSmntg 7gJ0lLcqgjfXtdu37wxj indexScreenshotToShow 1 available 2
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

I/flutter (31165): video -L7y4yAQAwN0f_YSmntg hIFyv3PUxXliHRwkKa5k indexScreenshotToShow 1 available 7
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

I/flutter (31165): video -L7y4yAQAwN0f_YSmntg 7gJ0lLcqgjfXtdu37wxj indexScreenshotToShow 0 available 2
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

I/flutter (31165): video -L7y4yAQAwN0f_YSmntg hIFyv3PUxXliHRwkKa5k indexScreenshotToShow 2 available 7
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 255
I/flutter (31165): image 216
I/flutter (31165): image 255
I/flutter (31165): image 224
I/flutter (31165): image 0
I/flutter (31165): image 16
I/flutter (31165): image 74
I/flutter (31165): image 70

I/flutter (31165): video -L7y4yAQAwN0f_YSmntg 7gJ0lLcqgjfXtdu37wxj indexScreenshotToShow 1 available 2
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 185
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): video -L7y4yAQAwN0f_YSmntg hIFyv3PUxXliHRwkKa5k indexScreenshotToShow 3 available 7
I/flutter (31165): image 147909
I/flutter (31165): image 1
I/flutter (31165): image 185
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
I/flutter (31165): image 0
E/flutter (31165): [ERROR:flutter/lib/ui/painting/codec.cc(89)] Failed decoding image. Data is either invalid, or it is encoded using an unsupported format.
I/flutter (31165): ══╡ EXCEPTION CAUGHT BY SERVICES ╞══════════════════════════════════════════════════════════════════
I/flutter (31165): The following _Exception was thrown resolving an image codec:
I/flutter (31165): Exception: operation failed
I/flutter (31165): ════════════════════════════════════════════════════════════════════════════════════════════════════
```


At one point, the Uint8List data started to change! Sure enough, it led to a codec error. Refer to "I/flutter (31165): video -L7y4yAQAwN0f_YSmntg 7gJ0lLcqgjfXtdu37wxj indexScreenshotToShow 1 available 2" - you will see it twice in log above. The first time, it has the expected data (as per log after Uint8List loaded from firebase) and the next time, it has different data!

I wrote all the code for the app and I was pretty sure I didn't manipulate the UintList data anywhere. Nonetheless, I checked that, and no, I do not manipulate that data. The app gets it from Firebase and
stores it in a variable. The variable is passed on to new Image.memory with fit: BoxFit.cover. This image is inside a container, of fixed width and height.

I should also note that I have some random crashes in the app since updating to Beta 3, such as Fatal signal 11 (SIGSEGV), code 1, fault addr 0x9df2800f in tid 31191 (1.ui)
Those always happen when the screenshots widget is building.
For example:
```
I/flutter (31165): video -L7y4yAQAwN0f_YSmntg hIFyv3PUxXliHRwkKa5k indexScreenshotToShow 3 available 7
I/flutter (31165): image 147909
I/flutter (31165): image 1
F/libc    (31165): Fatal signal 11 (SIGSEGV), code 1, fault addr 0x9df2800f in tid 31191 (1.ui)
```

All this happens on Android, I don't run it on iOS. It does happen on different versions of Android though (6.0 and 8.1).


So, in conclusion, my current thoughts are like this: 
1) either, I'm going crazy and need a rest,
2) or, in some cases, some of the Flutter engine/rendering code overwrites some of data in the Uint8List, which leads to codec error in some cases (most of them), and crashes (perhaps depending on timing of when the data is overwritten). 

I will try to create a simple app reproducing this issue, but I thought I would update my research on this.

**Comment by jason-simmons** [comment]
This sounds like the bug that was fixed by:
    https://github.com/flutter/engine/commit/b6be9377c58c488afea042a1b0ed8910b158c319

Try upgrading to the latest engine on the master branch.  The beta channel is currently pinned to a version that predates this fix.

**Comment by freewheelnat** [comment]
Unfortunately, I couldn't get my app to work on master channel.

I have a couple of IOExceptions from Firebase storage and it completely froze the app. I have the same IOExceptions when running app on Beta 3, but the app doesn't freeze (as all the code reading from Firebase is inside a `try{} catch(e){}` block). So I gave up, because it seems that master branch has broken something in my app, and as master branch isn't supposed to be stable anyway, I can't charge my client for time trying to make the app work in master branch.

Reading the code for the commit you mentioned, I do think you are correct, it will fix my issue. Do you have any rough timeline for when it will land in Beta, eg will it be in the next Beta? So I can update my client (our app is still in prototype/early testing stage).

**Comment by zoechi** [comment]
@spkingr 
Please add the output of `flutter doctor` and the versions of the packages you are using.

@freewheelnat is this still an issue?

**Comment by freewheelnat** [comment]
The issue is resolved for me, on dev branch (v0.5.5).

**Comment by zoechi** [comment]
@freewheelnat thanks a lot for the feedback.

**Comment by clarktank** [comment]
and here is a bit of my output error:
```
E/flutter (31165): [ERROR:flutter/lib/ui/painting/codec.cc(89)] Failed decoding image. Data is either invalid, or it is encoded using an unsupported format.
I/flutter (31165): ══╡ EXCEPTION CAUGHT BY SERVICES ╞══════════════════════════════════════════════════════════════════
I/flutter (31165): The following _Exception was thrown resolving an image codec:
I/flutter (31165): Exception: operation failed
I/flutter (31165):
I/flutter (31165): When the exception was thrown, this was the stack:
I/flutter (31165): #0      NetworkImage._loadAsync (package:flutter/src/painting/image_provider.dart:452:12)
I/flutter (31165): <asynchronous suspension>
I/flutter (31165): #1      NetworkImage.load (package:flutter/src/painting/image_provider.dart:425:14)
I/flutter (31165): #2      ImageProvider.resolve.<anonymous closure>.<anonymous closure> (package:flutter/src/painting/image_provider.dart:265:86)
I/flutter (31165): #3      ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:82:22)
I/flutter (31165): #4      ImageProvider.resolve.<anonymous closure> (package:flutter/src/painting/image_provider.dart:265:63)
I/flutter (31165): #5      SynchronousFuture.then (package:flutter/src/foundation/synchronous_future.dart:38:29)
I/flutter (31165): #6      ImageProvider.resolve (package:flutter/src/painting/image_provider.dart:263:30)
```

**Comment by clarktank** [comment]
I have solved the problem by using:

`urlTemplate: "https://api.mapbox.com/v4/{id}/{z}/{x}/{y}@2x.png?access_token=xxxxxx"`

instead of

`urlTemplate: "https://api.mapbox.com/styles/v1/xxxx"`

**Comment by Tennyson365** [comment]
utf-8 not supported?

**Comment by Tennyson365** [comment]
No PNG support?
